### PR TITLE
#269: fixed token class span output

### LINKED ISSUE #269: TE_TOKEN_CLASSIFICATION_UDF_WITH_SPAN misses text_data_span in the output
To map error messages to the input, we need to return also the text_data_span

https://github.com/exasol/transformers-extension/blob/594bf64035e69ffbe8c2d01f48625cd0ea6b64ea/exasol_transformers_extension/resources/templates/with_spans/span_token_classification_udf.jinja.sql#L13

### LINKED ISSUE #272: asserts in udf unit test ignore results
in our unit tests we use 
`    assert (
            OutputMatcher(actual_output, n_input_columns) == expected_output,
            mock_pipeline_factory.mock_calls == params.expected_model_counter)`

which is true if only one of the conditions is met. we need to use 

`    assert (
            OutputMatcher(actual_output, n_input_columns) == expected_output **and**
            mock_pipeline_factory.mock_calls == params.expected_model_counter)`

instead

### LINKED ISSUE #273: Use StandAloneUDFMock for unit test for token_classification_udf
- [ ] refactor unit tests for token_classification_udf to use StandAloneUDFMock
- [ ] delete everything that was made obsolete with this change
- [ ] refactor params files to be better maintainable

### PR DESCRIPTION
**Pull request by MarleneKress79789**
### All Submissions:

* [ ] Is the title of the Pull Request correct?
* [x] Is the title of the corresponding issue correct?
* [x] Have you updated the changelog?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../../pulls) for the same update/change?  
* [x] Are you mentioning the issue which this PullRequest fixes ("Fixes...")
* [x] Before you merge don't forget to run tests in AWS CodeBuild, by adding `[CodeBuild]` to the commit message

fixes #269 
fixes #272 
fixes #273

### CODE REVIEW

File: tests/unit_tests/udfs/test_token_classification.py [hunk 1]
```diff
@@ -1,57 +1,36 @@
...
 from tests.unit_tests.udf_wrapper_params.token_classification.single_model_single_batch_complete import \
     SingleModelSingleBatchComplete
 from tests.unit_tests.udf_wrapper_params.token_classification.single_model_single_batch_incomplete import \
     SingleModelSingleBatchIncomplete
+
 from tests.unit_tests.udfs.output_matcher import Output, OutputMatcher
 from tests.utils.mock_bucketfs_location import (fake_bucketfs_location_from_conn_object, fake_local_bucketfs_path)
+from tests.utils.mock_cast import mock_cast
 
+def udf_wrapper_empty():
```
Comment (tkilias) [inline tests/unit_tests/udfs/test_token_classification.py:27]: Not needed anymore for new version of mock

File: tests/unit_tests/udfs/test_token_classification.py [hunk 2]
```diff
@@ -114,92 +96,144 @@ def create_mock_metadata(udf_wrapper):
...
-        metadata=meta,
+    mock_exa = MockExaEnvironment(
+        metadata=mock_meta,
         connections=params.bfs_connections)
 
-    result = executor.run([Group(params.input_data)], exa)
-    result_output = Output(result[0].rows)
+    mock_base_model_factory: Union[ModelFactoryProtocol, MagicMock] = create_autospec(ModelFactoryProtocol,
+                                                                                      _name="mock_base_model_factory")
+    number_of_intendet_used_models = params.expected_model_counter# todo is this always same?
```
Comment (tkilias) [inline tests/unit_tests/udfs/test_token_classification.py:205]: This depends on the test case, so having the number in the parameters is correct

Comment (tkilias) [reply tests/unit_tests/udfs/test_token_classification.py:205]: Only the name expected_model_counters is a bit odd. Maybe expected_model_calls or something like thst

File: doc/changes/changes_2.2.0.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+- #272: Fixed unit tests assertions not working correctly
+
+### Documentation
+
+n/a
+
+### Refactorings
+
+- #273: refactored unit tests for token_classification_udf to use StandAloneUDFMock, made params files more maintainable
```
Comment (tkilias) [inline doc/changes/changes_2.2.0.md:23]:
```suggestion
- #273: Refactored unit tests for token_classification_udf to use StandAloneUDFMock, made params files more maintainable
```

File: tests/unit_tests/udf_wrapper_params/sequence_classification/error_on_prediction_single_model_multiple_batch.py [hunk 4]
```diff
@@ -52,8 +52,8 @@ class ErrorOnPredictionSingleModelMultipleBatch:
     """
     error on prediction, single model, multiple batch,
     """
-    expected_single_text_model_counter = 0
-    expected_text_pair_model_counter = 0
+    expected_single_text_model_counter = 1
+    expected_text_pair_model_counter = 1
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/sequence_classification/error_on_prediction_single_model_multiple_batch.py:56]: I guess, the value before was wrong

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py [hunk 5]
```diff
@@ -1,113 +1,54 @@
...
-
-    def run(ctx: UDFContext):
-        udf.run(ctx)
-
 class ErrorNotCachedMultipleModelMultipleBatch:
     """
     not cached error, multiple model, multiple batch
     """
-    expected_model_counter = 0
+    expected_model_counter = 1
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py:12]: Same as https://github.com/exasol/transformers-extension/pull/270/files#r1845243573 I guess

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py [hunk 6]
```diff
@@ -1,95 +1,36 @@
...
-
-    def run(ctx: UDFContext):
-        udf.run(ctx)
-
 class ErrorNotCachedSingleModelMultipleBatch:
     """
     not cached error, single model, multiple batch
     """
-    expected_model_counter = 1
+    expected_model_counter = 0
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py:12]: Same as https://github.com/exasol/transformers-extension/pull/270/files#r1845243573 I guess

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                              error_msg = error_msg):
+    entity_char_begin = text_start + token_start
+    entity_char_end = text_start + token_end
+    entity_docid = text_data_docid
+    return [(bucketfs_conn, sub_dir, model_name,
+             text_data_docid, text_data_char_begin, text_data_char_end, aggregation_strategy,
+             entity_covered_text, entity_type, score, entity_docid, entity_char_begin, entity_char_end,
+             error_msg)]
+
+#todo if use in all tests
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:63]: If it is a permanent todo, please create a ticket

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+agg_strategy_simple="simple"
+
+token="token"
+token_start = 2
+token_end = 4
+entity_type="ENTITY_TYPE"
+score=0.1
+error_msg = None
+
+#todo comment explain entity/token naming mess
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:21]: What is with this todo

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py (hunk 5 above)
Comment (tkilias) [reply tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py:12]: Improve docstring

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py (hunk 6 above)
Comment (tkilias) [reply tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py:12]: Improve doc string

File: tests/unit_tests/udf_wrapper_params/sequence_classification/error_on_prediction_single_model_multiple_batch.py (hunk 4 above)
Comment (tkilias) [reply tests/unit_tests/udf_wrapper_params/sequence_classification/error_on_prediction_single_model_multiple_batch.py:56]: improve docstrings

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py (hunk 5 above)
Comment (tkilias) [reply tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py:12]: improve docstrings

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py (hunk 6 above)
Comment (tkilias) [reply tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py:12]: Improve docstring

File: tests/unit_tests/udf_wrapper_params/token_classification/error_on_prediction_multiple_model_multiple_batch.py [hunk 9]
```diff
@@ -1,51 +1,10 @@
...
-        pipeline=params.mock_pipeline,
-        base_model=params.mock_factory,
-        tokenizer=MockSequenceTokenizer)
-
-    def run(ctx: UDFContext):
-        udf.run(ctx)
-
 class ErrorOnPredictionMultipleModelMultipleBatch:
     """
     not cached error, multiple model, multiple batch
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_on_prediction_multiple_model_multiple_batch.py:10]: improve docstring

File: doc/changes/changes_2.2.0.md (hunk 3 above)
Comment (tkilias) [inline doc/changes/changes_2.2.0.md:23]:
```suggestion
- #273: Refactored unit tests for token_classification_udf to use StandAloneUDFMock, made params files more maintainable
```

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py [hunk 10]
```diff
@@ -1,113 +1,54 @@
...
+
+    input_data = make_input_row(bucketfs_conn=bfs_conn1, sub_dir=subdir1, text_data=text1) * data_size + \
+                 make_input_row(bucketfs_conn=bfs_conn2, sub_dir=subdir2,
+                                model_name="non_existing_model",text_data=text2) * data_size
+
+    output_data = make_output_row(bucketfs_conn=bfs_conn1, sub_dir=subdir1, text_data=text1) * n_entities * data_size + \
+                  make_output_row(bucketfs_conn=bfs_conn2, sub_dir=subdir2,
+                                  model_name="non_existing_model",text_data=text2,
+                                  score=None, start=None, end=None, word=None, entity=None,
+                                  error_msg="Traceback") * 1 * data_size#error on load_model -> only one output per input
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py:29]:
```suggestion
                                  error_msg="Traceback") * 1 * data_size #error on load_model -> only one output per input
```

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py [hunk 11]
```diff
@@ -1,113 +1,54 @@
...
+
+    work_with_span_input_data = make_input_row_with_span(bucketfs_conn=bfs_conn1, sub_dir=subdir1,text_data=text1) * data_size + \
+                                make_input_row_with_span(bucketfs_conn=bfs_conn2, sub_dir=subdir2,
+                                                         model_name="non_existing_model",text_data=text2) * data_size
+
+    work_with_span_output_data1 = make_output_row_with_span(bucketfs_conn=bfs_conn1, sub_dir=subdir1) * n_entities * data_size
+    work_with_span_output_data2 = [(bfs_conn2, subdir2, "non_existing_model", text_docid, text_start, text_end, agg_strategy_simple,
+                                   None, None, None, text_docid, None, None,"Traceback")] * 1 * data_size#error on load_model -> only one output per input
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py:37]:
```suggestion
                                   None, None, None, text_docid, None, None,"Traceback")] * 1 * data_size #error on load_model -> only one output per input
```

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py [hunk 12]
```diff
@@ -1,95 +1,36 @@
...
-    token_docid = 1
-    start = 0
-    end = 20
 
-    token_start = 2
-    token_end = 4
+    input_data = make_input_row(model_name="non_existing_model") * data_size
+    output_data = make_output_row(model_name="non_existing_model",
+                                  score=None, start=None, end=None, word=None, entity=None,
+                                  error_msg="Traceback") * 1 * data_size#error on load_model -> only one output per input
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_single_model_multiple_batch.py:20]:
```suggestion
                                  error_msg="Traceback") * 1 * data_size #error on load_model -> only one output per input
```

File: tests/unit_tests/udf_wrapper_params/token_classification/error_on_prediction_multiple_model_multiple_batch.py [hunk 13]
```diff
@@ -54,60 +13,46 @@ class ErrorOnPredictionMultipleModelMultipleBatch:
...
-                  [("bfs_conn2", "sub_dir2", "model2", "error on pred",
-                    agg_strategy, None, None, None, None, None, "Traceback")
-                   ] * n_entities * data_size
+    tokenizer_model_output_df_model1 =  [make_model_output_for_one_input_row(number_entities=n_entities) * data_size]
+    tokenizer_model_output_df_model2_batch1 =  [make_model_output_for_one_input_row(number_entities=1, #error on pred -> only one output per input
+                                  score=None, start=None, end=None, word=None, entity_group=None,
+                                  )]
+    tokenizer_model_output_df_model2_batch2 =  [make_model_output_for_one_input_row(number_entities=1, #error on pred -> only one output per input
+                                  score=None, start=None, end=None, word=None, entity_group=None,
+                                  )] # do we expect error on prediction to happen once per input or once per entity? per input makes more sense right?
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_on_prediction_multiple_model_multiple_batch.py:46]: Yes, per Input makes more sense

File: tests/unit_tests/udf_wrapper_params/token_classification/error_on_prediction_single_model_multiple_batch.py [hunk 14]
```diff
@@ -1,52 +1,10 @@
...
-        pipeline=params.mock_pipeline,
-        base_model=params.mock_factory,
-        tokenizer=MockSequenceTokenizer)
-
-    def run(ctx: UDFContext):
-        udf.run(ctx)
-
 class ErrorOnPredictionSingleModelMultipleBatch:
     """
     error on prediction, single model, multiple batch,
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/error_on_prediction_single_model_multiple_batch.py:10]: Improve docstring

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+token_start = 2
+token_end = 4
+entity_type="ENTITY_TYPE"
+score=0.1
+error_msg = None
+
+def make_input_row(device_id=device_id, bucketfs_conn=bucketfs_conn, sub_dir=sub_dir,
+                   model_name=model_name, text_data=text_data, aggregation_strategy=agg_strategy_simple):
+    """
+    creates an input row for token classification without span usage as a list,
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:24]:
```suggestion
    Creates an input row for token classification without span usage as a list,
```

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    using base params for all params that are not specified.
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:25]:
```suggestion
    using default values for all parameters that are not specified.
```

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    the found token is called "word" in our non span udf output,
+    while the type/class of the found token is called "entity".
+    """
+    return [(bucketfs_conn, sub_dir, model_name,text_data, aggregation_strategy,
+             start,end,word, entity,score, error_msg)]
+
+def make_model_output_for_one_input_row(number_entities:int, aggregation_strategy=agg_strategy_simple,
+                                        entity_group=entity_type, score=score, word=token, start=token_start, end=token_end):
+    """
+    makes the output the model returns to the udf for one input row.
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:45]:
```suggestion
    Makes the output the model returns to the udf for one input row.
```

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    the found token is called "word" in the model output,
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:46]:
```suggestion
    The found token is called "word" in the model output,
```

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    while the type/class of the found token is called "entity_group".
+    unless aggregation_strategy == "none", then the type/class of the found
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:48]:
```suggestion
    Unless aggregation_strategy == "none", then the type/class of the found
```

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+agg_strategy_simple="simple"
+
+token="token"
+token_start = 2
+token_end = 4
+entity_type="ENTITY_TYPE"
+score=0.1
+error_msg = None
+
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:20]: Add type hints to functions

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+def make_model_output_for_one_input_row(number_entities:int, aggregation_strategy=agg_strategy_simple,
+                                        entity_group=entity_type, score=score, word=token, start=token_start, end=token_end):
+    """
+    makes the output the model returns to the udf for one input row.
+    the found token is called "word" in the model output,
+    while the type/class of the found token is called "entity_group".
+    unless aggregation_strategy == "none", then the type/class of the found
+    token is called "entity" in the model output.
+    returns a list of number_entities times the model output row.
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:50]: Use proper docstring formatting for return values, see other projects. I think, bucketfs has it.

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+        model_output_single_entities = {'entity_group': entity_group, 'score': score, 'word': word, 'start': start,
+                                        'end': end}
+    return [[model_output_single_entities] * number_entities]
+
+def make_input_row_with_span(device_id=device_id, bucketfs_conn=bucketfs_conn, sub_dir=sub_dir,
+                             model_name=model_name, text_data=text_data, text_data_docid=text_docid,
+                             text_data_char_begin=text_start, text_data_char_end=text_end,
+                             aggregation_strategy=agg_strategy_simple):
+    """
+    creates an input row for token classification with span usage as a list,
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:65]: See other commetst

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+             text_data_char_begin,text_data_char_end, aggregation_strategy)]
+
+def make_output_row_with_span(bucketfs_conn=bucketfs_conn, sub_dir=sub_dir,
+                              model_name=model_name, text_data_docid=text_docid, text_data_char_begin=text_start,
+                              text_data_char_end=text_end, aggregation_strategy=agg_strategy_simple,
+                              text_start=text_start, token_start=token_start, token_end =token_end,
+                              entity_covered_text=token, entity_type=entity_type, score=score,
+                              error_msg = error_msg):
+    """
+    creates an output row for token classification with span usage as a list,
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:78]: See other comments

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+    entity_char_end = text_start + token_end
+    entity_docid = text_data_docid
+    return [(bucketfs_conn, sub_dir, model_name,
+             text_data_docid, text_data_char_begin, text_data_char_end, aggregation_strategy,
+             entity_covered_text, entity_type, score, entity_docid, entity_char_begin, entity_char_end,
+             error_msg)]
+
+def make_number_of_strings(input_str: str, desired_number: int):
+    """
+    returns desired number of "input_strX", where X is counting up to desired_number.
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:92]: Docstring style

File: tests/unit_tests/udf_wrapper_params/token_classification/multiple_model_multiple_batch_incomplete.py [hunk 25]
```diff
@@ -54,60 +13,54 @@ class MultipleModelMultipleBatchIncomplete:
     batch_size = 3
     data_size = 2
     n_entities = 3
-    agg_strategy = "simple"
 
-    token_docid = 1
-    start = 0
-    end = 20
+    bfs_conn1, bfs_conn2 = make_number_of_strings(bucketfs_conn, 2) # todo why two in this test case? multiple model could still be same bfs con right?
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/multiple_model_multiple_batch_incomplete.py:17]: probably no specific reason might make sense to use the same bfsconn

File: tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py (hunk 5 above)
Comment (MarleneKress79789) [reply tests/unit_tests/udf_wrapper_params/token_classification/error_not_cached_multiple_model_multiple_batch.py:12]: new issue for docstring improvements #276

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py (hunk 7 above)
Comment (MarleneKress79789) [reply tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py]: removed

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py (hunk 21 above)
Comment (MarleneKress79789) [reply tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:50]: also added to new ticket

File: tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py (hunk 20 above)
Comment (MarleneKress79789) [reply tests/unit_tests/udf_wrapper_params/token_classification/make_data_row_functions.py:20]: also added to new ticket

File: tests/unit_tests/udf_wrapper_params/token_classification/multiple_model_multiple_batch_incomplete.py (hunk 25 above)
Comment (MarleneKress79789) [reply tests/unit_tests/udf_wrapper_params/token_classification/multiple_model_multiple_batch_incomplete.py]: changed

File: tests/unit_tests/udf_wrapper_params/token_classification/multiple_model_multiple_batch_complete.py [hunk 26]
```diff
@@ -13,34 +13,33 @@ class MultipleModelMultipleBatchComplete:
     data_size = 2
     n_entities = 3
 
-    bfs_conn1, bfs_conn2 = make_number_of_strings(bucketfs_conn, 2) # todo why two in this test case? multiple model could still be same bfs con right?
     sub_dir1, sub_dir2 = make_number_of_strings(sub_dir, 2)
```
Comment (tkilias) [inline tests/unit_tests/udf_wrapper_params/token_classification/multiple_model_multiple_batch_complete.py:16]: Maybe for another PR, also multiple subdir might be not necessary

**Review by tkilias — APPROVED** [review]

**Pull request merged**